src/server/scripts/Northrend/IcecrownCitadel/icecrown_citadel.cpp

```diff
@@ -1805,6 +1847,509 @@ struct npc_entrance_faction_leader : public ScriptedAI
+                        minion->AI()->DoZoneInCombat();
+            }
+        }
+        else if (action == ACTION_EVADE && _isInCombat)
+        {
+            _isInCombat = false;
+            // Update Darkfallens
+            Reset();
+        }
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        _scheduler.Update(diff);
+    }
+
+private:
+    TaskScheduler _scheduler;
+    GuidVector _minionGuids;
+    bool _isInCombat;
+};
+
+struct DarkFallenAI : public ScriptedAI
+{
+    DarkFallenAI(Creature* creature) : ScriptedAI(creature), isDoingEmotes(true), attackSpellId(0) { }
+
+    virtual void ScheduleSpells() = 0;
+
+    void Reset() override
+    {
+        scheduler.CancelAll();
+        scheduler.SetValidator([this]
+        {
+            return !me->HasUnitState(UNIT_STATE_CASTING);
+        })
+        .Schedule(1s, 10s, [this](TaskContext emote)
+        {
+            if (roll_chance_i(20))
+            {
+                std::vector<Creature*> creatures;
+                Trinity::CreatureListSearcher<MinionSearch> searcher(me, creatures, MinionSearch(true));
+                Cell::VisitGridObjects(me, searcher, 10.0f);
+                if (!creatures.empty())
+                {
+                    Creature* friendly = Trinity::Containers::SelectRandomContainerElement(creatures);
+                    DoCast(friendly, SPELL_POLYMORPH_ALLY);
+                }
+            }
+            scheduler.Schedule(1s, [this](TaskContext /*emote*/)
+            {
+                me->HandleEmoteCommand(Trinity::Containers::SelectRandomContainerElement(DarkFallensEmotes));
+            });
+            emote.Repeat(15s, 30s);
+        });
+        isDoingEmotes = true;
+    }
+
+    void JustEngagedWith(Unit* who) override
+    {
+        isDoingEmotes = false;
+        scheduler.CancelAll();
+        ScheduleSpells();
+        if (Unit* trigger = ObjectAccessor::GetUnit(*me, triggerGuid))
+            trigger->GetAI()->DoAction(ACTION_COMBAT);
+    }
+
+    void DoAction(int32 action) override
+    {
+        if (action == ACTION_SIPHON_INTERRUPTED)
+            if (GameObject* orb = me->FindNearestGameObject(GO_EMPOWERING_BLOOD_ORB, 10.0f))
+                orb->RemoveFlag(GAMEOBJECT_FLAGS, GO_FLAG_NOT_SELECTABLE);
+    }
+
+    void SetGUID(ObjectGuid const& guid, int32 id) override
+    {
+        if (id == DATA_GUID)
+            triggerGuid = guid;
+    }
+
+    void JustReachedHome() override
+    {
+        ScriptedAI::JustReachedHome();
+        if (Unit* trigger = ObjectAccessor::GetUnit(*me, triggerGuid))
+            trigger->GetAI()->DoAction(ACTION_EVADE);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim() && !isDoingEmotes)
+            return;
+
+        scheduler.Update(diff);
+
+        if (attackSpellId)
+            DoSpellAttackIfReady(attackSpellId);
+        else
+            DoMeleeAttackIfReady();
+    }
+
+protected:
+    TaskScheduler scheduler;
```

Comment:
Protected members should be UpperCamelCase according to C++ Development Standards in TC wiki. (Had to look it up myself tbh, but lowerCamelCase looks way too local scope likely)

Reply:
i dont know that, will update.